Volume QA part II: fixes and tests

apostasie opened the pull request:
This is part 2 of #3125 - same idea.

The original motivation was to significantly increase testing coverage & quality, convergence with docker behavior, and stability wrt concurrency for anything related to volume lifecycle.

In the process of doing so, several issues cropped-up, that this PR also fix.

Tests:
- parallelized (which does also serve as a very good way to uncover raciness)
- expanded
- cleaned-up / normalized
- added namespace testing
- test more thoroughly, both in terms of cases and validating content of stdout / stderr
- ~~move tests from _linux to all platforms (not sure why it was linux only in the first place)~~ - wish I had a windows env to test

Code changes:
- docker align: "no labels" should be `nil`, not `[]string{}`
- docker align: `--label ""` should error
- docker align: multi-inspect that is partly failing, and other partly failing conditions
- remove useless / duplicated code (specifically around validation of identifiers that was performed multiple times)
- make sure all methods of the volume store are locking (to avoid racy behavior)
- replace broken/racy logic "`get` then `create`" by just `create` (since behavior has been recently changed and now always returns success)
- make the underlying filesystem implementation private (eg: no longer expose `Path`) to prevent leaking implementation to the consumer (see for example `cmd/nerdctl/namespace.go b/cmd/nerdctl/namespace.go` that was reading the filesystem directly instead of calling `List`)
- added a "persistent" `Lock` and `Unlock` methods on the store, for operations that need to call multiple methods successively or need guarantees that volumes cannot be modified at all until they are done
- cleaned-up and expanded `lockutil`, so that we don't expose anymore `Flock` (unimplemented for non Unix), and instead expose new lock/unlock methods
- reviewed all the codebase for any cases where that global lock call is necessary and added them
- reviewed codebase to fix #3132 (and fix #3123 hopefully)

While this is a significant PR in term of apparent code change size, the added testing should give us a good confidence boost on what's happening wrt volumes.

apostasie commented:
@AkihiroSuda what's the timeline for v2?

Do I still have time for more or are we frozen for v2 already?

I would have liked to cleanup / fix all bugs for `login` overall, and maybe some network work, but there is only so much I can do today - so, if you intend on freezing, lmk.

AkihiroSuda commented:
> v2

Planned to be synced with containerd v2
https://github.com/orgs/containerd/projects/9

apostasie commented:
> > v2
> 
> Planned to be synced with containerd v2 https://github.com/orgs/containerd/projects/9

Cool.

So I guess we also need this in then: https://github.com/containerd/stargz-snapshotter/pull/1545

And #2902 to pass.

Lmk if there is something I can do to help with ^.

AkihiroSuda commented:
nerdctl v2 will have containerd v2 binaries in the `nerdctl-full` bundle, but I guess it is not necessary to use Go packages of containerd v2

> Lmk if there is something I can do to help with ^.

Thanks, I'd appreciate if you can help #2902 and its dependencies (stargz, soci, etc.) 🙇

apostasie commented on cmd/nerdctl/volume_create.go:
```diff
@@ -45,6 +48,12 @@ func processVolumeCreateOptions(cmd *cobra.Command) (types.VolumeCreateOptions,
 	if err != nil {
 		return types.VolumeCreateOptions{}, err
 	}
+	for _, label := range labels {
```
Align with docker behavior. Empty labels are not allowed.

apostasie commented on cmd/nerdctl/namespace.go:
```diff
@@ -113,11 +113,11 @@ func namespaceLsAction(cmd *cobra.Command, args []string) error {
 			log.L.Warn(err)
 		} else {
-			volEnts, err := os.ReadDir(volStore)
+			volEnts, err := volStore.List(false)
```
Replace direct access to underlying filesystem implementation by a proper api call.

apostasie commented on pkg/cmd/container/create.go:
```diff
@@ -151,7 +164,7 @@ func Create(ctx context.Context, client *containerd.Client, args []string, netMa
 
 	var mountOpts []oci.SpecOpts
-	mountOpts, internalLabels.anonVolumes, internalLabels.mountPoints, err = generateMountOpts(ctx, client, ensuredImage, options)
+	mountOpts, internalLabels.anonVolumes, internalLabels.mountPoints, err = generateMountOpts(ctx, client, ensuredImage, volStore, options)
```
Since we are now locking the volume store, we need to pass it along.

apostasie commented on pkg/cmd/volume/create.go:
```diff
@@ -32,9 +31,6 @@ func Create(name string, options types.VolumeCreateOptions) (*native.Volume, err
 	}
-	if err := identifiers.Validate(name); err != nil {
-		return nil, fmt.Errorf("malformed name %s: %w", name, err)
-	}
```
Useless, as this is performed by the volume store.

apostasie commented on pkg/cmd/volume/inspect.go:
```diff
@@ -17,23 +17,40 @@
 		var vol, err = volStore.Get(name, options.Size)
 		if err != nil {
-			return err
+			warns = append(warns, err)
```
Align with docker behavior

apostasie commented on pkg/cmd/volume/prune.go:
```diff
@@ -26,29 +26,47 @@ import (
 	}
-	usedVolumes, err := usedVolumes(ctx, containers)
+
+	usedVolumesList, err := usedVolumes(ctx, containers)
```
Better variable name...

apostasie commented on pkg/cmd/volume/prune.go:
```diff
@@ -26,29 +26,47 @@ import (
 			continue
 		}
 		if !options.All {
+			if volume.Labels == nil {
```
Now that labels can be nil, need to check for that.

apostasie commented on pkg/cmd/volume/rm.go:
```diff
@@ -32,47 +32,61 @@ import (
 	for _, name := range volumes {
-		if _, ok := usedVolumes[name]; ok {
-			return fmt.Errorf("volume %q is in use", name)
+		if _, ok := usedVolumesList[name]; ok {
```
Align with docker behavior

apostasie commented on pkg/composer/up_volume.go:
```diff
@@ -41,6 +41,9 @@ func (c *Composer) upVolume(ctx context.Context, shortName string) error {
+	// FIXME: this is racy. By the time we get below to creating the volume, there is no guarantee that things are still fine
+	// Furthermore, volStore.Get no longer errors if the volume already exists (docker behavior), so, the purpose of this
+	// call needs to be assessed (it might still error if the name is malformed, or if there is a filesystem error)
 	volExists, err := c.VolumeExists(fullName)
```
I do not understand the compose code. Why do we shell out?
Leaving these untouched for now.

AkihiroSuda replied:
> shell out

IIRC it was just for ease of quick implementation

apostasie commented on pkg/lockutil/lockutil_unix.go:
```diff
@@ -32,18 +32,18 @@ func WithDirLock(dir string, fn func() error) error {
 }
 
-func Flock(f *os.File, flags int) error {
+func flock(f *os.File, flags int) error {
```
This should not be public. It is unix only.

apostasie commented on pkg/mountutil/mountutil.go:
```diff
@@ -203,16 +201,11 @@ func handleAnonymousVolumes(s string, volStore volumestore.VolumeStore) (volumeS
-	vol, err := volStore.Get(res.Name, false)
+
+	// Create returns an existing volume or creates a new one if necessary.
+	vol, err := volStore.Create(res.Name, nil)
```
No reason for any of that anymore.

apostasie commented on pkg/mountutil/mountutilmock/volumestore.mock.go:
```diff
@@ -106,16 +106,26 @@ func (m *MockVolumeStoreMockRecorder) Remove(names any) *gomock.Call {
 
-// Dir mocks the Dir method of VolumeStore
-func (m *MockVolumeStore) Dir() string {
+// Lock mocks the Lock method of VolumeStore
```
The mock stuff. Still have no idea what is this and what to do here.
Merely ensuring we fullfill the interface here, but if you have info on what I should do here, please speak.

apostasie commented on pkg/mountutil/volumestore/volumestore.go:
```diff
@@ -31,21 +31,33 @@ import (
 )
 
-// Path returns a string like `/var/lib/nerdctl/1935db59/volumes/default`.
-func Path(dataStore, ns string) (string, error) {
```
This should not be public.
Consumer should only use the interface methods and not fiddle with the underlying (filesystem based) implementation.

apostasie commented on pkg/mountutil/volumestore/volumestore.go:
```diff
@@ -55,175 +67,246 @@ func New(dataStore, ns string) (VolumeStore, error) {
+	return nil
 }
 
-func (vs *volumeStore) Dir() string {
```
See other comment on publicizing these.

apostasie commented on cmd/nerdctl/system_prune_linux_test.go:
```diff
@@ -32,6 +32,11 @@ import (
 
 func TestSystemPrune(t *testing.T) {
 	testutil.RequiresBuild(t)
+	// FIXME: for some reason, using a dedicated namespace does work well with rootless, but fails with rootful... :/
```
No idea what's going on here, and why this happens only with rootful.
Will investigate later when I get to system tests...

apostasie commented on cmd/nerdctl/volume_prune_linux_test.go:
```diff
@@ -17,30 +17,137 @@
+		t.Parallel()
+	}
+
+	// FIXME: for an unknown reason, when testing ipv6, calling NewBaseWithNamespace per sub-test, in the tearDown/tearUp methods
```
No idea.........
I guess it is another thing to add to the list for whomever is going to cleanup our testing helpers eventually.

apostasie commented:
@AkihiroSuda this is finally ready - sorry it took so long to get there.

Also it is bigger that I wanted it to be - sorry about that as well. I guess problems just kept on piling up as I parallelized more...

Pure code changes are relatively small and targeted.
The only really new / interesting bit is the introduction of the "durable" locking mechanism for `VolumeStore` that allow operations like `container create` or `container remove` to lock the store for the duration of their operation.

The rest is merely small bug fixes - and a boatload of new tests.

I have added a bunch of github comments alongside the code to facilitate review. 

PTAL at your convenience.

Thanks!

AkihiroSuda commented on cmd/nerdctl/volume_create.go:
```diff
@@ -17,6 +17,9 @@
 import (
+	"fmt"
+
+	"github.com/containerd/containerd/errdefs"
```
Yes, but probably it should be a separate PR

apostasie replied:
@djdongjin I do not think this will work, unless we update everything (including dependencies) to v2, as both packages will yield different results wrt `errors.Is`

apostasie replied:
Correction: may actually work with 1.7.19 - still, certainly better as a separate PR.

AkihiroSuda commented on cmd/nerdctl/system_prune_linux_test.go:
```diff
@@ -32,6 +32,11 @@ import (
 func TestSystemPrune(t *testing.T) {
 	testutil.RequiresBuild(t)
+	// FIXME: for some reason, using a dedicated namespace does work well with rootless, but fails with rootful... :/
+	// Maybe something related to buildctl/kit?
```
Probably because BuildKit is not enabled for rootless CI?

AkihiroSuda commented on cmd/nerdctl/volume_prune_linux_test.go:
```diff
@@ -17,30 +17,137 @@
+		//base.Cmd("volume", "rm", "-f", tID+"-busy").Run()
+		// base.Cmd("volume", "rm", "-f", tID+"-free").Run()
+		// base.Cmd("volume", "rm", "-f", anonIDFree).Run()
+		// base.Cmd("volume", "rm", "-f", anonIDBusy).Run()
```
Is this WIP?

apostasie replied:
No. This should have been cleaned-up - it just escaped me.

Let me do one last go through the diff and make sure there is no leftovers.

Thanks for catching it.

apostasie commented:
Rebased.

@AkihiroSuda last tiny push does remove the spurious commented leftovers, extra line, and updated the comment on rootful/buidkit.

Just did another proofread of the diff -  we should be good to go (let's make sure the CI is happy of course).

AkihiroSuda reviewed (approved):
Thanks

apostasie commented:
> Thanks

Thanks a lot @AkihiroSuda 
Appreciate your time and effort reviewing these.
